### DISCUSSION TITLE
Dynamic Service Subscriber

### DISCUSSION THREAD

**Discussion by jdsaphir** [symfony/symfony#49877] [Q&A] [answered]
Hi everyone! I'm building an app that will let users trigger commands. The available commands are stored in a database and `$command->getClass()` returns the FQCN of the class that contains the command's logic.

I've created a Service Subscriber following the [documentation's example](https://symfony.com/doc/current/service_container/service_subscribers_locators.html#defining-a-service-subscriber), and was quite amazed to see it work. However, as the getSubscribedServices() method is static, I had to list the command classes in the code manually.

What I would like is for the list of command classes to be pulled directly from the database, so that I can add a command just by creating the class file and an entry in the DB, so something like that:
```PHP
public function getSubscribedServices(): array
{
    foreach($this->em->getRepository(ChatCommand::class)->findAll() as $command)
        $subscribedServices[] = $command->getClass();

    return $subscribedServices;
}
```

That's my first time using a Service Subscriber, so honestly I'm not even sure that's the right solution for what I'm trying to do, any advice would be greatly appreciated!

**Comment by ghost** [comment]
This is not doable with static service subscriber. If you will create a class file anyway then what's the cost of adding an entry to getSubscribedServices() method? You have to change the code anyway.
However what you want is to inject the whole container or (better practice) create a service locator (smaller container, for limited type of services) for your commands.

**Reply by jdsaphir** [reply]
I don't want to have to edit another file than the new command's class, and as the list is already in the database, it doesn't feel clean having to manually maintain a second list in the code.
I've looked at service locators but I think I would run into the same issue of having to list the classes to be added to the locator manually?

**Reply by ghost** [reply]
It looks like you want to have to much magic out of the box.

> I've looked at service locators but I think I would run into the same issue of having to list the classes to be added to the locator manually?

Use `AutoconfigureTag` attribute on top of you command class or create a custom marker interface and configure it to tag all the instances with a given tag. Maybe you can use a custom attribute instead of the marker interface - I don't know if Symfony allows that. Then configure the $container argument as tagged locator of specific tag.

**Comment by stof** [comment, marked as answer]
This looks more like a use case for `ServiceLocatorArgument` (which can be autowired by using the `TaggedLocator` attribute) than for `ServiceSubscriber` here.

**Reply by jdsaphir** [reply]
Hi! Thank you so much for your answer! Took me a while to understand what you were talking about, but once I did, it was stupidly easy!

**Reply by ghost** [reply]
This is exactly what my comments say.

**Reply by jdsaphir** [reply]
Indeed and I apologize, I didn't understand it at the time.
